Support member functions with qualifiers in differentiation calls

parth-07 opened the pull request:
This pull request :
-  Add support of functions with qualifiers (`const`, `volatile`,`reference qualifiers` and `noexcept`)
    in differentiation calls using:
   - `clad::differentiate`
   - `clad::gradient`
   - `clad::hessian`
   - `clad::jacobian`
- Add support of vardiac functions for forward differentiation mode (`clad::differentiation`)
  
Consequently, it fix differentiation of member functions with qualifiers in forward differentiation mode .

Main idea used in this PR is suggested by @vgvassilev and @pdimov

Fixes #199

parth-07 commented:
@vgvassilev Suggested changes implemented. Now,
- `clad::differentiate`
- `clad::gradient`
- `clad::hessian`
- `clad::jacobian`

each consist of just one definition(no overloading), same definition works for both member and non-member functions. And all member function types and qualifiers are now supported. 

Added regression tests for forward differentiation of following function types and function reference qualifiers
- ordinary member functions
- const member functions
- reference qualifiers (&, const&, &&, const&&)
- volatile member functions
- noexcept and const noexcept member functions

I **have not** added regression tests for member functions for reverse, hessian, and jacobian mode, since differentiating class member functions for these modes give compile time error. #139

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -52,43 +52,120 @@ namespace clad {
   private:
     CladFunctionType m_Function;
-    char* m_Code;
+    char *m_Code;
```
Can you undo changes of that kind?

parth-07 replied:
Done, just undid these changes.

parth-07 commented:
@vgvassilev Please review the PR.

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -52,37 +52,98 @@ namespace clad {
+  // ExtractDerivedFnTraits is used to get type of the derived functions 
+  // in reverse, hessian and jacobian differentiation modes
+  template<class ReturnType>
+  struct ExtractDerivedFnTraits {};
```
We will need the same specializations for `ExtractDerivedFnTraits` as for `return_type` (kudos @pdimov).

parth-07 replied:
Yes, but since, derivation of member functions are currently not working for reverse, jacobian, and hessian modes(The modes which use extractDerivedFnTraits) . I did not add these specializatons,I will add them soon.

vgvassilev replied:
A `FIXME:` explaining that should be fine. Also, you could add a test with a declaration of the functions (not their definitions), expect clad to complain but also test the specializations...

parth-07 replied:
I will add the specialization, and also explain the situation briefly in comments. 
And yes, will add these tests then :)

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -127,6 +189,7 @@ namespace clad {
     }
   };
 
+
```
No new line here.

parth-07 replied:
okay.

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -108,15 +169,16 @@ namespace clad {
-        // static_cast == std::forward, i.e convert the Args to ArgTypes
-        ::execute(m_Function, static_cast<ArgTypes>(args)...);
+
+      // return CallType<std::is_member_function_pointer<decltype(
```
Do we need this comment?

parth-07 replied:
No, it was left by mistake.

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -52,37 +52,98 @@ namespace clad {
+  }
+
+
+  // Using std::function and std::mem_fn introduces a lot of overhead, which we
```
Duplicated comment?

parth-07 replied:
Removed it.

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -52,37 +52,98 @@ namespace clad {
-    using type = ReturnResult (*)(ArgTypes...);
+  // Trait class to deduce return type of function at commpile time
+  // Only function pointer types are supported by this trait class
+  template <class F> struct return_type {};
```
We seem to be missing `noexcept &` specializations. So in general we need 2*3*2 specializations if we do not support `volatile`: `const` or not, multiplied times `&`, `&&`, or nothing, times `noexcept` or not (for c++17 and above). cc: @pdimov

parth-07 replied:
noexcept with reference qualifiers (noexcept &, noexcept &&) is giving error for both C++11 and C++14 in clang. So I am leaving out noexcept with reference qualifiers specializations, I will add comment there that we may need these specializations for later C++ versions.

So, if we support `volatile` , then on considering, `const` or `non-const`, `volatile` or `non-volatile`, reference qualifiers(`&` , `&&`, no reference qualifier) will give 2 * 2 * 3 specializations, and `noexcept`, `const noexcept`, `volatile noexcept` and `const volatile noexcept` will give 4 more. 
And then if we also consider functions with variable argument list(variadic functions) for all these functions, we will have total of **2 * (2 * 2 * 3 + 4) = 32 specializations.** 
I will try to take cover all the specializations this time, sorry for missing out so many of them before.

parth-07 replied:
Until C++17, `noexcept` is not a part of function type. So, we can safely remove all specializations involving `noexcept`.

parth-07 replied:
noexcept with reference qualifier is not giving error, I was using wrong syntax, apparently noexcept should be placed after reference qualifiers (`& noexcept` , `&& noexcept` etc).

vgvassilev commented on include/clad/Differentiator/Differentiator.h:
```diff
@@ -52,37 +52,98 @@ namespace clad {
 
+  // for executing member-functions
+  template<class ReturnType, class C, class Obj, class... Args> auto\
+  execute_helper( ReturnType C::* f, Obj&& obj, Args&&... args) -> return_type_t<decltype(f)>  {
```
Can you make sure that this works in C++11 mode and add an explicit test for compiling this header file in c++11 mode?

parth-07 replied:
Yes, this works in C++11. I always used `-std=c++11`flag while testing. I will add the explicit test file.

vgvassilev replied:
Hm, okay. And you did not get any warnings that this is a non-c++11 feature?

parth-07 replied:
For testing, using `lit` we are already compiling all the test files with `-std=c++11` flag. Do I need to add explicit test some other way ?

vgvassilev replied:
Ha, fair point, an oversight on my end!

parth-07 replied:
@vgvassilev no, there were no warnings of that kind,and I just rechecked, this feature was introduced in C++11.

vgvassilev replied:
Nice, every day I learn something new. Thanks!

parth-07 commented:
@vgvassilev Now, there are **26** specializations for `return_type` trait class and **13** specializations for `ExtractDerivedFnTraits` trait class, I have not included specializations for variadic functions for `ExtractDerivedFnTraits` class since then position of `*_result` pointer in derived function will be ambigious.

I would like to propose putting trait classes(`return_type` and `ExtractDerivedFnTraits`) in a separate header file, it will increase code modularity and maintainability.

vgvassilev commented:
> @vgvassilev Now, there are **26** specializations for `return_type` trait class and **13** specializations for `ExtractDerivedFnTraits` trait class, I have not included specializations for variadic functions for `ExtractDerivedFnTraits` class since then position of `*_result` pointer in derived function will be ambigious.

Yes, that is a bug in our implementation. Could you open a new issue for this so we can keep track of it?

> I would like to propose putting trait classes(`return_type` and `ExtractDerivedFnTraits`) in a separate header file,as it will increase code modularity and maintainability.

Yes, makes sense to me.

parth-07 commented:
> > @vgvassilev Now, there are **26** specializations for `return_type` trait class and **13** specializations for `ExtractDerivedFnTraits` trait class, I have not included specializations for variadic functions for `ExtractDerivedFnTraits` class since then position of `*_result` pointer in derived function will be ambigious.
> 
> Yes, that is a bug in our implementation. Could you open a new issue for this so we can keep track of it?

Yes, sure.

> > I would like to propose putting trait classes(`return_type` and `ExtractDerivedFnTraits`) in a separate header file,as it will increase code modularity and maintainability.
> 
> Yes, makes sense to me.

Do you have suggestion for the file name ? How about `FunctionTraits.h` or just `Traits.h` ?

vgvassilev commented:
> Do you have suggestion for the file name ? How about `FunctionTraits.h` or just `Traits.h` ?

I'd prefer `FunctionTraits.h`

parth-07 commented:
> > Do you have suggestion for the file name ? How about `FunctionTraits.h` or just `Traits.h` ?
> 
> I'd prefer `FunctionTraits.h`

Okay :) 

And one last question, should I include specialization involving `noexcept` ? They will only be needed for C++17 and beyond, currently I have added `FIXME`explaining they will be required later on.

vgvassilev commented:
> > > Do you have suggestion for the file name ? How about `FunctionTraits.h` or just `Traits.h` ?
> > 
> > 
> > I'd prefer `FunctionTraits.h`
> 
> Okay :)
> 
> And one last question, should I include specialization involving `noexcept` ? They will only be needed for C++17 and beyond, currently I have added `FIXME`explaining they will be required later on.

Yes, you should, they should be conditional on if clang was called with `-std=c++17` -- we had that condition in the godbolt link I believe. Then you will probably need to add two more `//RUN:` clauses in your test compiling with `-std=c++14` and `-std=c++17` (and perhaps `-std=c++1z` conditionally on the clang version).

parth-07 commented:
Added specializations for all permutations of (const or non-const), (volatile or non-volatile), (&, &&, no-reference-qualifier), (noexcept, no-noexcept) and additional vardiac functions for `return_type`.
Now there are **50** specializations of `return_type` trait class, and **25** specializations of `ExtractDerivedFnTraits` trait class. 

I have also added additional `RUN` clauses in tests for also testing with `c++14` and `c++17`. Do I additionally need to add `RUN` clauses for `-std=c++1z` ? Isn't it same as `-std=c++17` ?

vgvassilev reviewed (commented):
Overall looks good. Is there a reason for not squashing all commits into one?

I'd suggest improving the commit message style. Maybe [that](https://chris.beams.io/posts/git-commit/) would be a useful read.

parth-07 commented:
> Overall looks good. Is there a reason for not squashing all commits into one?

No, there wasn't any particular reason for not squashing all commits into one. I have now squashed all commits into one commit.

> I'd suggest improving the commit message style. Maybe [that](https://chris.beams.io/posts/git-commit/) would be a useful read.

Yes, the link provided was very useful, I have also tried to improve commit message style in the squash commit.

vgvassilev reviewed (approved):
Thanks for working on this! LGTM!
